### ISSUE TITLE
'count' in elasticsearch has some problem

### ISSUE THREAD

**Issue by dreamingo88** [Impetus/kundera#843]
Entity:
`
public class Category {

```
@Id
public String id;

@Required
@MinSize(3)
public String name;

@Required
@MinSize(3)
public String code;
```

}
`

Create 50 rows of data:
`
    for(int i = 1; i <= 50; i++) {
            Category ctg = new Category();
            ctg.id = ""+i; 
            ctg.name = UUID.randomUUID().toString; 
            ctg.code = "abc"; 
            em.persist(ctg); 
        }
`

Count 1:
`
    Query q = em.createQuery("select count(ctg) from Category ctg where ctg.code = 'abc'");
    String count= q.getSingleResult().toString();
    System.out.println("count="+count);
`
print: count=50, works right.

Count 2:
`
    Query q = em.createQuery("select count(ctg.name) from Category ctg where ctg.code = 'abc'");
    String count= q.getSingleResult().toString();
    System.out.println("count="+count);
`
print: count=801, **wrong!!!**

remove all datas, and create another 50 rows of data like this:
`
    for(int i = 1; i <= 50; i++) {
            Category ctg = new Category(); 
            ctg.id = UUID.randomUUID().toString;
            ctg.name = 'test name ' + i;
            ctg.code = "abc";
            em.persist(ctg);
        }
`

Count 3:
`
    Query q = em.createQuery("select count(ctg) from Category ctg where ctg.code = 'abc'");
    String count= q.getSingleResult().toString();
    System.out.println("count="+count);
`
print: count=787, **wrong!!!**

Count 4:
`
    Query q = em.createQuery("select count(ctg.name) from Category ctg where ctg.code = 'abc'");
    String count= q.getSingleResult().toString();
    System.out.println("count="+count);
`
print: count=150, **wrong!!!**

Please check and fix it, thanks!

**Comment by karthikprasad13** [comment]
@dreamingo88 

Hey,

Are you starting elasticsearch from Kundera (embedded mode) or deployed externally? Also, How are you deleting the data from elasticsearch.. manually by deleting data directory or through Kundera?

-Karthik

**Comment by dreamingo88** [comment]
@karthikprasad13 
Hi,
I deployed externally. 
I deleting the data from elasticsearch (with plugin head).

**Comment by dreamingo88** [comment]
Is there any news about this problem?

**Comment by devender-yadav** [comment]
@dreamingo88 

We are almost done with Kundera-Elasticsearch update (_Elasticsearch 2.3_). We will try to replicate this issue  with this latest version.

-Dev

**Comment by devender-yadav** [comment]
@dreamingo88 

Elasticsearch update is added to the trunk. Can you please build code from the trunk and try to replicate your issues now.

You can clone source code using : 

```
git clone https://github.com/impetus-opensource/Kundera.git
```

-Dev

**Comment by devender-yadav** [comment]
@dreamingo88 

Could you please try this with Kundera 3.5

-Dev

**Comment by dreamingo88** [comment]
@devender-yadav 
Sorry for delayed reply.
I will try soon.

**Comment by dreamingo88** [comment]
@devender-yadav 
Hi, I have test it today, It works right.
Thank you!
You can close this issue.

**Comment by devender-yadav** [comment]
@dreamingo88 

By default, indexes in elasticsearch are not updated immediately after insertion of data. So count query just after insertion of data may lead to wrong results.

We added a property to update indexes immediately. If  user has a usecase to search on newly added data immediately after addition of it, this property can be used. 

```
em.setProperty("es.refresh.indexes", true);
```

https://github.com/impetus-opensource/Kundera/issues/882 is related issue.

Check [testcase](https://github.com/impetus-opensource/Kundera/blob/c645d7f7e949a0e9f9b06b010e527725a4b0a3d2/src/kundera-elastic-search/src/test/java/com/impetus/client/es/CRUDTestWithRefreshedIndexes.java) for more details. 

This will be available in Kundera 3.6 (_to be released in  next week_).

-Dev
